hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -688,6 +665,35 @@ public OmKeyInfo lookupKey(OmKeyArgs args, String clientAddress)
     }
   }
 
+  /**
+   * Refresh pipeline info in OM by asking SCM.
+   * @param value OmKeyInfo
+   */
+  @VisibleForTesting
+  protected void refreshPipeline(OmKeyInfo value) throws  IOException {
+    Map<Long, ContainerWithPipeline> containerWithPipelineMap = new HashMap<>();
+    for (OmKeyLocationInfoGroup key : value.getKeyLocationVersions()) {
+      for (OmKeyLocationInfo k : key.getLocationList()) {
+        // TODO: fix Some tests that may not initialize container client
+        // The production should always have containerClient initialized.
+        if (scmClient.getContainerClient() != null) {
+          if (containerWithPipelineMap.containsKey(k.getContainerID())) {
+            ContainerWithPipeline containerWithPipeline = scmClient
+                .getContainerClient()
+                .getContainerWithPipeline(k.getContainerID());
```

Comment:
Here it throws ContainerNotFoundException, can we convert that to OMException.

As in ozoneManagerRequestHandler.java L453, if it is anything other than OMException we throw it as INTERNAL_ERROR.
if (ex instanceof OMException) {
      return Status.values()[((OMException) ex).getResult().ordinal()];
    } else {
      if (LOG.isDebugEnabled()) {
        LOG.debug("Unknown error occurs", ex);
      }
      return Status.INTERNAL_ERROR;
    }

Reply:
I don't mind changing it. But, if the SCM logs it as ContainerNotFound Exception in logs, we may be OK since in a way we are actually not able to "find" the key here?

Reply:
If we can throw OMException with some new code/message it will be clearer. And it also helps when we see this kind of error.

Reply:
Sure, I will change this. SCM getContainerWithPipeline API can throw a number of exceptions like PipelineNotFound, ContainerNotFound, SCMException etc. I can make a generic "SCM_EXCEPTION" code in OMException.

Reply:
Added new OMException ResultCode.